pkg/apis/autoscaling/types.go

```diff
@@ -271,6 +297,31 @@ type ResourceMetricSource struct {
 	Target MetricTarget
 }
 
+// PodResourceMetricSource indicates how to scale on a resource metric known to
+// Kubernetes, as specified in requests and limits, describing each pod in the
+// current scale target (e.g. CPU or memory).  The values will be averaged
+// together before being compared to the target.  Such metrics are built in to
+// Kubernetes, and have special scaling options on top of those available to
+// normal per-pod metrics using the "pods" source.  Only one "target" type
+// should be set.
+type PodResourceMetricSource ResourceMetricSource
+
+// ContainerResourceMetricSource indicates how to scale on a resource metric known to
+// Kubernetes, as specified in the requests and limits, describing a single container in
+// each of the pods of the current scale target(e.g. CPU or memory). The values will be
+// averaged together before being compared to the target. Such metrics are built into
+// Kubernetes, and have special scaling options on top of those available to
+// normal per-pod metrics using the "pods" source. Only one "target" type
+// should be set.
+type ContainerResourceMetricSource struct {
+	// name is the name of the of the resource
+	Name api.ResourceName `json:"name"`
+	// container is the name of the container in the pods of the scaling target.
+	Container string `json:"container"`
```

Comment:
this adds the possibility of several edge cases:
1. the named container do not exist in some (or all) of the selected pods
1. the named container is defined differently in some of the selected pods (as an initContainer, or without requests, or with requests)

need to document and test behavior for these scenarios

also need to explain how someone is expected to modify HPA when adding/removing/modifying a container within a deployment... HPA first, then deployment? deployment first, then HPA?

Reply:
I looked into this and the controller handles this case when metrics are missing for a pod. It then fills pod's metrics cautiously. What that means is that it recomputes the recommendation with the missing pod metrics as 0 if the calculation indicates that the deployment should scale up, and as 100% utilization if the previous calculation shows that the target should be scaled down. This was added to handle the case where metrics might be missing due to broken metrics reporting on certain nodes. I'm not sure if the same behavior should be used when only some pods have the container name. 

Do you think it is sufficient if we just add a disclaimer that all the targetted pods should have the container with that name.

Also for deployment rollouts, I think it is sufficient if both the old and new container names are part of the HPA spec and after the rollout only the new container name is kept. This should be prevent unexpected scale downs.

Let me know what you think @liggitt 

cc @josephburnett

Reply:
Currently HPA treats missing metrics and missing request slightly differently.

If any container of a pod is missing metrics, the HPA will treat the whole pod as missing metrics: https://github.com/kubernetes/kubernetes/blob/62b8faed4042e9f1a9c1e895ad1c36bb473eb6ee/pkg/controller/podautoscaler/metrics/rest_metrics_client.go#L80-L89  In this case, as @arjunrn pointed out, we will make conservative assumptions about the metrics and proceed with autoscaling.

However if any container is missing a resource request for a configured target, HPA will refuse to take any action.  This is considered an invalid configuration (no way to calculate utilization without a denominator) so the HPA fails without taking any action.

If the user provides a container target for a container that doesn't exist, it would be more natural to treat this as an invalid configuration and do nothing.  Once the container shows up, reconciliation will proceed.  E.g. to add a new container, one should add it to the deployment and wait for all replica sets to be updated.  Then update the HPA.  In this way, there will be no period during which autoscaling is off.  However, if the user adds the container to HPA first, then the deployment, autoscaling will be suspended until all pods have the new container.  Which is not ideal, but it will stabilize to a working configuration once the deployment is finished rolling out.